File: src/main/java/io/github/mg138/tsbook/entities/effect/EffectHandler.java
package io.github.mg138.tsbook.entities.effect;

import io.github.mg138.tsbook.Book;
import io.github.mg138.tsbook.entities.effect.data.EntityStatusEffect;
import io.github.mg138.tsbook.entities.effect.data.map.RegisteredEffects;
import io.github.mg138.tsbook.entities.effect.data.StatusEffect;
import io.github.mg138.tsbook.entities.effect.data.StatusEffectType;
import org.bukkit.entity.Entity;
import org.bukkit.entity.LivingEntity;
import org.bukkit.event.Listener;
import org.bukkit.scheduler.BukkitRunnable;

import java.util.*;

public class EffectHandler implements Listener {
    private final static HashMap<LivingEntity, HashMap<StatusEffectType, BukkitRunnable>> activeRunnable = new HashMap<>();
    private final static HashMap<LivingEntity, HashMap<StatusEffectType, StatusEffect>> activeEffect = new HashMap<>();

    public static void applyRawEffect(Entity target, StatusEffect effect, BukkitRunnable runnable, int delay, int period) {
        if (!(target instanceof LivingEntity)) return;

        addRunnable((LivingEntity) target, effect, runnable);
        addEffect((LivingEntity) target, effect);
        runnable.runTaskTimer(Book.inst, delay, period);
    }

    public static void apply(StatusEffectType type, Entity target, double power, int time) {
        if (!(target instanceof LivingEntity)) return;

        if (activeRunnable.containsKey(target)) {
            HashMap<StatusEffectType, BukkitRunnable> runnables = activeRunnable.get(target);
            if (runnables.containsKey(type)) {
                runnables.get(type).cancel();
                runnables.remove(type);
            }
        }

        if (activeEffect.containsKey(target)) {
            activeEffect.get(target).remove(type);
        }

        RegisteredEffects.Effects.get(type).accept(new EntityStatusEffect(target, type, power, time));
    }

    public static void unload() {
        HashSet<BukkitRunnable> removing = new HashSet<>();
        activeRunnable.forEach((entity, runnableMap) ->
                runnableMap.forEach((type, runnable) ->
                        removing.add(runnable)
                )
        );
        removing.forEach(BukkitRunnable::cancel);
        activeRunnable.clear();
        activeEffect.clear();
    }

    public static void addEffect(LivingEntity entity, StatusEffect effect) {
        HashMap<StatusEffectType, StatusEffect> effects = activeEffect.getOrDefault(entity, new HashMap<>());

        effects.put(effect.type, effect);
        activeEffect.put(entity, effects);
    }

    public static StatusEffect getEffect(LivingEntity entity, StatusEffectType type) {
        HashMap<StatusEffectType, StatusEffect> effects = getEffects(entity);
        if (effects == null) return null;

        return effects.get(type);
    }

    public static HashMap<StatusEffectType, StatusEffect> getEffects(LivingEntity entity) {
        return activeEffect.get(entity);
    }

    public static boolean hasEffect(LivingEntity entity, StatusEffectType type) {
        HashMap<StatusEffectType, StatusEffect> effects = getEffects(entity);
        if (effects == null) return false;

        return effects.containsKey(type);
    }

    public static void addRunnable(LivingEntity entity, StatusEffect effect, BukkitRunnable runnable) {
        HashMap<StatusEffectType, BukkitRunnable> runnables = activeRunnable.getOrDefault(entity, new HashMap<>());

        BukkitRunnable old = runnables.put(effect.type, runnable);
        if (old != null) old.cancel();

        activeRunnable.put(entity, runnables);
    }

    public static void remove(LivingEntity entity, StatusEffectType type) {
        if (!activeEffect.containsKey(entity) || !activeRunnable.containsKey(entity)) return;
        activeEffect.get(entity).remove(type);
        activeRunnable.get(entity).remove(type);
    }

    public static void remove(LivingEntity entity) {
        HashMap<StatusEffectType, BukkitRunnable> removing;
        try {
            removing = new HashMap<>(activeRunnable.get(entity));
        } catch (NullPointerException e) {
            throw new NullPointerException("The entity doesn't have any active effects");
        }
        removing.forEach((type, runnable) -> runnable.cancel());
        removing.clear();
        activeRunnable.remove(entity);
        activeEffect.remove(entity);
    }
}

File: src/main/java/io/github/mg138/tsbook/entities/effect/EffectHandler.kt
package io.github.mg138.tsbook.entities.effect

import io.github.mg138.tsbook.Book
import org.bukkit.entity.LivingEntity
import io.github.mg138.tsbook.entities.effect.data.StatusEffectType
import org.bukkit.scheduler.BukkitRunnable
import io.github.mg138.tsbook.entities.effect.data.StatusEffect
import io.github.mg138.tsbook.entities.effect.data.map.RegisteredEffects
import io.github.mg138.tsbook.entities.effect.data.EntityStatusEffect
import org.bukkit.entity.Entity
import java.lang.NullPointerException
import java.util.*
import kotlin.collections.HashMap

object EffectHandler {
    private val activeRunnable: MutableMap<LivingEntity, MutableMap<StatusEffectType, BukkitRunnable>> = HashMap()
    private val activeEffect: MutableMap<LivingEntity, MutableMap<StatusEffectType, StatusEffect>> = HashMap()

    fun applyRawEffect(target: Entity, effect: StatusEffect, runnable: BukkitRunnable, delay: Long, period: Long) {
        if (target !is LivingEntity) return
        addRunnable(target, effect, runnable)
        addEffect(target, effect)
        runnable.runTaskTimer(Book.inst, delay, period)
    }

    fun apply(type: StatusEffectType, target: Entity, power: Double, time: Long) {
        if (target !is LivingEntity) return
        activeRunnable[target]?.get(type)?.cancel()
        activeEffect[target]?.remove(type)

        RegisteredEffects.Effects[type]?.invoke(EntityStatusEffect(target, type, power, time))
    }

    fun unload() {
        val removing: MutableSet<BukkitRunnable> = HashSet()
        activeRunnable.forEach { (_, runnableMap) ->
            runnableMap.forEach { (_, runnable) ->
                removing.add(runnable)
            }
        }
        removing.forEach { it.cancel() }
        removing.clear()
        activeRunnable.clear()
        activeEffect.clear()
    }

    fun getEffect(entity: LivingEntity, type: StatusEffectType): StatusEffect? {
        val effects = activeEffect[entity] ?: return null
        return effects[type]
    }

    fun hasEffect(entity: LivingEntity, type: StatusEffectType): Boolean {
        val effects = activeEffect[entity] ?: return false
        return effects.containsKey(type)
    }

    private fun addEffect(entity: LivingEntity, effect: StatusEffect) {
        activeEffect.putIfAbsent(entity, EnumMap(StatusEffectType::class.java))
        activeEffect[entity]?.set(effect.type, effect)
    }

    private fun addRunnable(entity: LivingEntity, effect: StatusEffect, runnable: BukkitRunnable) {
        val runnables = activeRunnable.getOrDefault(entity, EnumMap(StatusEffectType::class.java))
        runnables.put(effect.type, runnable)?.cancel()
        activeRunnable[entity] = runnables
    }

    fun remove(entity: LivingEntity, type: StatusEffectType) {
        activeEffect[entity]?.remove(type)
        activeRunnable[entity]?.get(type)?.cancel()
    }

    fun remove(entity: LivingEntity) {
        val removing = try {
            activeRunnable[entity]!!
        } catch (e: NullPointerException) {
            throw NullPointerException("The entity doesn't have any active effects")
        }
        removing.forEach { (_, runnable) -> runnable.cancel() }
        removing.clear()
        activeRunnable.remove(entity)
        activeEffect.remove(entity)
    }
}